File: src/main/java/com/keygenqt/mylibrary/models/Book.java
/*
 * Copyright 2020 Vitaliy Zarubin
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.keygenqt.mylibrary.models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.NotFound;
import org.hibernate.annotations.NotFoundAction;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.util.Date;
import java.util.List;

@Entity
@Table(name = "books")
public class Book {

    public final static String COVER_SOFT = "Soft";
    public final static String COVER_SOLID = "Solid";
    public final static String COVER_OTHER = "Other";

    public Book() {
    }

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Column(name = "genre_id", nullable = false)
    private Long genreId;

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Column(name = "image")
    private String image;

    @Column(name = "title")
    private String title;

    @Column(name = "author")
    private String author;

    @Column(name = "publisher")
    private String publisher;

    @Column(name = "isbn")
    private String isbn;

    @Column(name = "year")
    private Integer year;

    @Column(name = "number_of_pages")
    private Integer numberOfPages;

    @Column(name = "description")
    private String description;

    @Column(name = "cover_type", nullable = false)
    private String coverType;

    @Column(name = "sale", nullable = false)
    private Boolean sale = false;

    @JsonIgnore
    @Column(name = "enabled", nullable = false)
    private Boolean enabled = true;

    @JsonIgnore
    @CreationTimestamp
    @Column(name = "created_at")
    @Temporal(TemporalType.TIMESTAMP)
    private Date createdAt;

    @JsonIgnore
    @UpdateTimestamp
    @Column(name = "updated_at")
    @Temporal(TemporalType.TIMESTAMP)
    private Date updatedAt;

    @OneToOne(cascade = CascadeType.ALL)
    @NotFound(action = NotFoundAction.IGNORE)
    @JoinColumn(name = "genre_id", referencedColumnName = "id", updatable = false, insertable = false)
    private Genre genre;

    @OneToOne(cascade = CascadeType.ALL)
    @NotFound(action = NotFoundAction.IGNORE)
    @JoinColumn(name = "user_id", referencedColumnName = "id", updatable = false, insertable = false)
    private User user;

    public Book(Long genreId,
                Long userId,
                String image,
                String title,
                String author,
                String publisher,
                String isbn,
                String year,
                String numberOfPages,
                String description,
                String coverType,
                Boolean sale) {
        this.genreId = genreId;
        this.userId = userId;
        this.image = image;
        this.title = title;
        this.author = author;
        this.publisher = publisher;
        this.isbn = isbn;
        this.year = year == null || year.isEmpty() ? null : Integer.parseInt(year);
        this.numberOfPages = numberOfPages == null || numberOfPages.isEmpty() ? null : Integer.parseInt(numberOfPages);
        this.description = description;
        this.coverType = coverType == null ? Book.COVER_OTHER : coverType;
        this.sale = sale;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Long getGenreId() {
        return genreId;
    }

    public void setGenreId(Long genreId) {
        this.genreId = genreId;
    }

    public Long getUserId() {
        return userId;
    }

    public void setUserId(Long userId) {
        this.userId = userId;
    }

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getPublisher() {
        return publisher;
    }

    public void setPublisher(String publisher) {
        this.publisher = publisher;
    }

    public String getIsbn() {
        return isbn;
    }

    public void setIsbn(String isbn) {
        this.isbn = isbn;
    }

    public Integer getYear() {
        return year;
    }

    public void setYear(String year) {
        this.year = year == null || year.isEmpty() ? null : Integer.parseInt(year);
    }

    public Integer getNumberOfPages() {
        return numberOfPages;
    }

    public void setNumberOfPages(String numberOfPages) {
        this.numberOfPages = numberOfPages == null || numberOfPages.isEmpty() ? null : Integer.parseInt(numberOfPages);
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getCoverType() {
        return coverType;
    }

    public void setCoverType(String coverType) {
        this.coverType = coverType;
    }

    public Boolean getSale() {
        return sale;
    }

    public void setSale(Boolean sale) {
        this.sale = sale;
    }

    public Boolean getEnabled() {
        return enabled;
    }

    public void setEnabled(Boolean enabled) {
        this.enabled = enabled;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Date createdAt) {
        this.createdAt = createdAt;
    }

    public Date getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Date updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Genre getGenre() {
        return genre;
    }

    public void setGenre(Genre genre) {
        this.genre = genre;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }
}

File: src/main/kotlin/com/keygenqt/mylibrary/models/Book.kt
///*
// * Copyright 2020 Vitaliy Zarubin
// *
// * Licensed under the Apache License, Version 2.0 (the "License");
// * you may not use this file except in compliance with the License.
// * You may obtain a copy of the License at
// *
// *     http://www.apache.org/licenses/LICENSE-2.0
// *
// * Unless required by applicable law or agreed to in writing, software
// * distributed under the License is distributed on an "AS IS" BASIS,
// * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// * See the License for the specific language governing permissions and
// * limitations under the License.
// */
//
//package com.keygenqt.mylibrary.models
//
//import com.fasterxml.jackson.annotation.*
//import org.hibernate.annotations.*
//import java.util.*
//import javax.persistence.*
//import javax.persistence.CascadeType
//import javax.persistence.Entity
//import javax.persistence.Table
//
//@Entity
//@Table(name = "books")
//data class Book(
//    @Id
//    @GeneratedValue(strategy = GenerationType.IDENTITY)
//    var id: Long? = null,
//
//    @Column(name = "genre_id", nullable = false)
//    var genreId: Long = 0,
//
//    @Column(name = "user_id", nullable = false)
//    var userId: Long = 0,
//
//    @Column(name = "image", nullable = true)
//    var image: String? = null,
//
//    @Column(name = "title", nullable = true)
//    var title: String? = null,
//
//    @Column(name = "author", nullable = true)
//    var author: String? = null,
//
//    @Column(name = "publisher", nullable = true)
//    var publisher: String? = null,
//
//    @Column(name = "isbn", nullable = true)
//    var ISBN: String? = null,
//
//    @Column(name = "year", nullable = true)
//    var year: Int? = null,
//
//    @Column(name = "number_of_pages", nullable = true)
//    var numberOfPages: Int? = null,
//
//    @Column(name = "description", nullable = true)
//    var description: String? = null,
//
//    @Column(name = "cover_type", nullable = false)
//    var coverType: String = "soft",
//
//    @Column(name = "sale", nullable = false)
//    var sale: Boolean = false,
//
//    @Column(name = "enabled", nullable = false)
//    var enabled: Boolean = true,
//
//    @JsonIgnore
//    @CreationTimestamp
//    @Column(name = "created_at")
//    @Temporal(TemporalType.TIMESTAMP)
//    var createdAt: Date = Date(),
//
//    @JsonIgnore
//    @UpdateTimestamp
//    @Column(name = "updated_at")
//    @Temporal(TemporalType.TIMESTAMP)
//    var updatedAt: Date = Date(),
//
//    @OneToOne(cascade = [CascadeType.ALL])
//    @NotFound(action = NotFoundAction.IGNORE)
//    @JoinColumn(name = "genre_id", referencedColumnName = "id", updatable = false, insertable = false)
//    var genre: Genre? = null,
//
//    @OneToOne(cascade = [CascadeType.ALL])
//    @NotFound(action = NotFoundAction.IGNORE)
//    @JoinColumn(name = "user_id", referencedColumnName = "id", updatable = false, insertable = false)
//    var user: User? = null
//) {
//    companion object {
//        const val COVER_SOFT = "Soft"
//        const val COVER_SOLID = "Solid"
//        const val COVER_OTHER = "Other"
//    }
//}